Add GitHub guide

vladimir-v-diaz opened the pull request:
@lukpueh and I thought it made more sense to have the Git+GitHub guide available on the lab-guidelines repo.  This pull request addresses issue secure-systems-lab/code-style-guidelines/issues/13

vladimir-v-diaz committed:
Add git+GitHub doc

vladimir-v-diaz committed:
List the Git+GitHub guide in README

aaaaalbert committed:
Make Wikipedia and GitHub URLs use HTTPS

aaaaalbert commented on git-github-introduction.md:
````diff
@@ -0,0 +1,413 @@
+You can check your user information by typing the following command:
+
+```
+$ git config --list
````
If you've used Git already on this machine, this command yields a long scary list of config items.

For checking just the user name and email, one could use `git config --get-regexp user`

aaaaalbert commented on git-github-introduction.md:
````diff
@@ -0,0 +1,413 @@
+configuration if you use the global flag (as in the provided example).
+
+```
+$ git config --global user.name "John Smith"
````
An irrelevant BTW: Configuring `user.name` seems to be optional, only the email address appears to be a mandatory config item. Also, the user name need not be your actual name.

vladimir-v-diaz replied:
I noted in the guide that the `user.name` value is arbitrary, but it is customary to use your full name.  I also decided to retain the user.name configuration since the official Git docs advise users to set it, and the error message one receives when a Git identity is unset asks for a `user.name`.

https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup

aaaaalbert committed:
s/SeattleTestbed/Secure Systems Lab/

aaaaalbert committed:
Clarify sentence about commands for local copy

aaaaalbert committed:
Note that `pull` is a Git (not GitHub) command

aaaaalbert commented on git-github-introduction.md:
```diff
@@ -0,0 +1,414 @@
+Fetching and merging the latest changes of `upstream` (remembering that the
+`upstream` alias points to the original repository) can be done with Git's
+`pull` command.  The `pull` command fetches changes made to the remote
+repository and merges them into your local repository.  Branches are multiple
```
I suggest to break the paragraph and add a new subsection, "Branching".

aaaaalbert commented on git-github-introduction.md:
```diff
@@ -0,0 +1,414 @@
+![GitHub Branches](/images/branches_sm.png)
+
+Additional information on [syncing a
+fork](https://help.github.com/articles/syncing-a-fork/) is recommended.
```
"recommended" as in "a recommended read"? Should this be, `[The GitHub help](...) has more information on keeping your forks in sync with upstream`?

aaaaalbert committed:
Replace reference to Seattle with "upstream repo"

Then point out that we're talking about the *example* again.

aaaaalbert committed:
Mention "New PR" button label, not (wrong) color

aaaaalbert committed:
Add code-style formatting for merge conflicts

aaaaalbert committed:
Clarify stashing

(Not that I use it much myself, but I think that this is how it should be used.)

aaaaalbert committed:
Fix filename of Git guide in README

aaaaalbert committed:
Add link to choosealicense.com for MIT license

The site is run by GitHub and presents a nice overview of license "features".

aaaaalbert reviewed (commented):
I read through the document, clicked all of the links, fixed a few trivial issues, and left a couple of comments. Feel free to address as you see fit.

aaaaalbert commented:
How can we help  our newcomers to adopt Git quickly (to use it productively for their own work) and in a way that interfaces nicely with our workflows?

Should we send newcomers to @octocat's `Spoon-Knife` repo to actually try out the theoretic skills acquired through this guide?

Or shall we have a playground repo of our own where people can experiment? "First task on first day on the team: Add your favorite sports/treat/metro line to secure-systems-lab/playground" sorts of thing?

baloian commented:
How about something like this
https://codingsec.net/2016/06/frequently-used-git-commands/

vladimir-v-diaz committed:
Edit section for Git identity

* Make it clear that the user.name is arbitrary
* Change title from 'Configure User information' to 'Configure Git Identity'

vladimir-v-diaz committed:
Indicate that your configuration settings are printed with 'git config --list'

vladimir-v-diaz committed:
Break up paragraph and add a 'Branch' subsection

vladimir-v-diaz committed:
Link to GitHub's 'Git Cheat Sheet'

vladimir-v-diaz commented:
@baloyan 
We used to have a bunch of those Git Cheat Sheets in the old lab.  I've added a link to it in the Guide, since I found it helpful in the past.
https://services.github.com/on-demand/downloads/github-git-cheat-sheet.pdf

vladimir-v-diaz commented:
@aaaaalbert 
Thanks for the excellent review and edits.

aaaaalbert commented:
Thanks for addressing my concerns, @vladimir-v-diaz!
